Add ability to save failed score

cdwcgt opened the pull request:
Related to https://github.com/ppy/osu/issues/8721 and https://github.com/ppy/osu/discussions/18770

F rank now is -1 in `ScoreRank` and will changing in the future

maybe https://github.com/ppy/osu/pull/10339 's design is better than this pr.
- [x] Consider copying this design

[Not now](https://github.com/ppy/osu/pull/18785#discussion_r901876139)
~~`[LocalisableDescription(typeof(BeatmapsStrings), nameof(BeatmapsStrings.RankF))]`~~
~~we have no Localisable String for RankF~~
~~https://github.com/ppy/osu-resources/blob/master/osu.Game.Resources/Localisation/Web/BeatmapsStrings.cs~~
~~- [ ] add Localisable String in osu-resources (if necessary)~~

the score processing for failed score has some problem (like 99% F score with sudden death)
But maybe deal this problem future?

cdwcgt committed:
Save score button on failed screen

cdwcgt committed:
Let F to -1 temporary

cdwcgt committed:
deal with test

bdach commented on osu.Game/Scoring/ScoreRank.cs:
```diff
@@ -11,6 +11,11 @@ namespace osu.Game.Scoring
 {
     public enum ScoreRank
     {
+        // TODO: reconsider changing later on
+        [LocalisableDescription(typeof(BeatmapsStrings), nameof(BeatmapsStrings.RankD))]
```
Why is this even added? I checked source, `LocalisableDescription` [will gracefully go back to `Description` if the localisable variant isn't found](https://github.com/ppy/osu-framework/blob/2d4142c5433492868f1ba7a2f670afeb84a9e906/osu.Framework/Extensions/ExtensionMethods.cs#L197-L199). So not adding this attribute at all would have been less wrong than doing this.

cdwcgt replied:
I'm sorry that I didn't read the framwork code carefully.
So just remove this attribute and wait for RankF string added to osu.Resource?

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -1043,6 +1044,22 @@ public override bool OnExiting(ScreenExitEvent e)
+
+            try
+            {
+                ImportScore(scoreCopy).ConfigureAwait(false);
```
Why is this async call fire-and-forgotten? No `await`, no `.ContinueWith()` even?

cdwcgt replied:
Because I am considering the return type of this method.
I use void for it but code quality said that avoid use void return type for async methods

So it's a bit like my stupid solution.

I will revert it and consider return type

bdach commented on osu.Game/Scoring/ScoreRank.cs:
```diff
@@ -11,6 +11,11 @@ namespace osu.Game.Scoring
+        // TODO: reconsider changing later on
+        [LocalisableDescription(typeof(BeatmapsStrings), nameof(BeatmapsStrings.RankD))]
+        [Description(@"F")]
+        F = -1,
```
Why go to the trouble of adding this rank *if you're not even using it for the feature it was intended for*? I just checked this branch, the score imports with rank D - so what was the point?

cdwcgt replied:
Maybe I commit it to my test branch, I remembered that I use `Score,ScoreInfo.Rank = F` at saveReplay()
I will fix it

peppy commented:
Zero test coverage means you likely did not read the contribution guide. You should be developing using visual tests, not the game itself.

cdwcgt commented:
It looks like I opened this pr too hurry.
I will try to deal with all test error this week.

cdwcgt committed:
Use better design and fix some problem

Let saveReplay async but still void
Make failed score's rank = F

cdwcgt committed:
Shake button when replay already save

cdwcgt committed:
Remove useless `LocalisableDescription`

cdwcgt committed:
Add button state, fix async issues, watch replay method

Most borrowed from `ReplayDownloadButton`

cdwcgt commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,141 @@
+                    break;
+
+                case ImportState.Failed:
+                    button.State.Value = DownloadState.NotDownloaded;
```
```suggestion
```
Forget to delete,
If everything is ok I will delete it

peppy replied:
I don't understand. If you forgot, could you please commit a fix for it?

peppy replied:
Please push the changes. We can't review incomplete changes.

cdwcgt committed:
remove useless value change

peppy committed:
Merge branch 'master' into Save-Score-Failed

cdwcgt commented:
> @cdwcgt you requested a review, but there are no tests yet. Are you planning to add any, or is that for us to do?

I may have misunderstood the meaning of the test
I thought you meant code quality testing

I will try to add test unit for it

cdwcgt commented:
> @cdwcgt you requested a review, but there are no tests yet. Are you planning to add any, or is that for us to do?

I may have misunderstood the meaning of the test
I thought you meant code quality testing

I will try to add test unit for it

peppy commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,140 @@
+            Scheduler.Add(saveScoreDelegate);
+        }
+
+        private void updateTooltip(ValueChangedEvent<ImportState> state)
+        {
+            switch (state.NewValue)
+            {
+                case ImportState.Imported:
+                    button.TooltipText = @"Watch replay";
+                    break;
+
+                case ImportState.Importing:
+                    button.TooltipText = @"Importing score";
+                    break;
+
+                case ImportState.Failed:
+                    button.TooltipText = @"Import failed, click button to re-import";
+                    break;
+
+                default:
+                    button.TooltipText = @"Save score";
+                    break;
+            }
+        }
```
I don't know about this. We've never used a tooltip to show progress of an ongoing operation.

cdwcgt replied:
I borrowed from [here](https://github.com/ppy/osu/pull/10339/files#diff-82cb1b3ae319044e80814e92987d37d68782835752177cbab90f610eca2bbad5R31-R43)

https://github.com/ppy/osu/blob/master/osu.Game/Beatmaps/Drawables/Cards/Buttons/DownloadButton.cs#L55

I will rename it to `updateState()`

by the way it should be possible to do localization here, but it seems not now?

peppy replied:
Localisation is still a bit of an afterthought, probably fine to leave for this PR (there's enough to fix already).

peppy commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,140 @@
+            }
+        }
+
+        public enum ImportState
```
I don't think this should be required.

cdwcgt replied:
What I originally intended to use was `DownloadState`
but DownloadState doesn't have `ImportFailed` State so I create a new enum here

in fact `DownloadState` need a `failed` stat?
The original button will revert to `NotDownloaded` when the download fails

peppy replied:
What's the case where it can fail?

cdwcgt replied:
May only have a hard disk write issue in the current version and may be able to be resolved by re-importing
in past versions (before https://github.com/ppy/osu/pull/19025), custom rulesets also fail to import, and repeating the import doesn't help

peppy commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,140 @@
+            State.Value = ImportState.Importing;
+            saveReplayAsync = Task.Run(SaveReplay);
+
+            saveScoreDelegate = new ScheduledDelegate(() =>
```
Use `ContinueWith` on the task rather than this weird looping logic.

peppy committed:
Rename lots of weird variables

peppy commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,140 @@
+                        break;
+
+                    case ImportState.Importing:
+                        shakeContainer.Shake();
```
I'd prefer the button was disabled during import, not shaking.

peppy commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,140 @@
+                switch (State.Value)
+                {
+                    case ImportState.Imported:
+                        game?.PresentScore(score, ScorePresentType.Gameplay);
```
Dunno about this. Import shouldn't be required to watch a scores. maybe okay for now but the UX is pretty shoddy.

cdwcgt replied:
so it should delete it and make button disabled after imported?

peppy commented on osu.Game/Screens/Play/SaveFailedScoreButton.cs:
```diff
@@ -0,0 +1,140 @@
+
+        private void saveScore()
+        {
+            State.Value = ImportState.Importing;
```
Assert that this method can never be run if the task is already non-null.

cdwcgt committed:
Use ContinueWith, Check is Task empty

cdwcgt committed:
disabled button instead of shake

removed `ShakeContainer`

peppy commented:
For testing, please wait for https://github.com/ppy/osu/pull/19025 to be merged as it contains a new test scene which should work great for this purpose.

cdwcgt committed:
rename updateTooltip to updateState

nagi-desuuu reviewed (commented):
I might also bring up a feature in osu!stable that can also export replays of a failed song.
I think the hot key for that was F2, if I still remember.

peppy commented:
@cdwcgt are you interested in writing a test for this one? i'll probably go through it today in a hope of getting it in the next release (depending on how the UX goes). if preferred i can add the test.

cdwcgt commented:
> are you interested in writing a test for this one? i'll probably go through it today in a hope of getting it in the next release (depending on how the UX goes). if preferred i can add the test.

I might have some difficulty writing tests because I'm trying to understand the specific process
If you can help me write it, I'll gladly accept it.

Thanks

peppy committed:
Merge branch 'master' into Save-Score-Failed

peppy committed:
Tidy up implementation and remove unnecessary enum

peppy committed:
Simplify bindable logic greatly

peppy committed:
Make delegate firing more safe to being set later than BDL

peppy commented:
I've tidied the code up and two points remain:

- Watching a replay doesn't... fail... it just plays forever. Feels pretty bad?
- Tests (will add soon)

peppy committed:
Add test coverage of saving a failed score

peppy committed:
Check whether score is already imported and show correct state for save button

peppy committed:
Allow failed scores to fail in replay playback

peppy commented:
I've added test coverage and fixed watching a failed replay not failing.

@smoogipoo can you take a look over this (UX mostly) and check you're okay with it? It's not perfect but a good start and fine to get in IMO.

frenzibyte committed:
Move default state steps to `SetUp` rather than duplicating per test case

frenzibyte committed:
Merge branch 'master' into Save-Score-Failed

frenzibyte committed:
Fix failed scores not prepared before import

frenzibyte committed:
Update save score button to check availability after import

Previously was relying on whether `SaveReplay` returns null, but since
I've changed it to use the standard "prepare score for import" path, the
button has to check for local availability after import since that path
doesn't return null on fail.
